### PR TITLE
TILA-1606 | Standardized graphql query optimization

### PR DESCRIPTION
**Pull request by Niilis**
Standardized utils function to optimize GraphQL queries dynamically.

Applied query optimizer for applications field as an example and testing field.

### CODE REVIEW

File: utils/query_performance.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+        # TODO: Currently, this optimization only supports queries with one node
+        # If multiple nodes are added to query, this implementation does not
+        # know how to optimize the n:th query!
+
+        # Start parsing GraphQL query structure
+        for selection in operation.selection_set.selections:
+            if selection.name.value == field_name:
+                query_selections = selection.selection_set.selections
+
```
Comment (nicobav) [inline utils/query_performance.py:57]: I'd get rid off the air here

File: utils/query_performance.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                break
+
+        # Query selections found, find edges
+        if query_selections:
+            for selection in query_selections:
+                if selection.name.value == "edges":
+                    edge_selections = selection.selection_set.selections
+
```
Comment (nicobav) [inline utils/query_performance.py:65]: I'd get rid off the air here

Comment (nicobav) [inline utils/query_performance.py:65]: I'd get rid off the air here

**Review by nicobav — CHANGES_REQUESTED** [review]
Looks great :+1: 

Only some line breaks to get rid off :shrug:

**Review by nicobav — APPROVED** [review]
amazing!

**Review by saaste — APPROVED** [review]
Took me a while to understand all the bells and whistles and I still don't quite comprehend everything - mostly because of my tiny Django experience. But the code is easy to read and follow and in general it makes sense.

So, LGTM!

I wonder if something like this has been already done and provided as a package. This feels such a common problem that one would think that we can't been the only one struggling with query optimization 😄

**Comment by Niilis** [comment]
> Took me a while to understand all the bells and whistles and I still don't quite comprehend everything - mostly because of my tiny Django experience. But the code is easy to read and follow and in general it makes sense.
> 
> So, LGTM!
> 
> I wonder if something like this has been already done and provided as a package. This feels such a common problem that one would think that we can't been the only one struggling with query optimization smile

@saaste there is this one [Graphene Django query optimizer](https://github.com/tfoxy/graphene-django-optimizer), but I did not get it to work on deeper levels than 1-2. Within this project, some structures are many levels more deep and will definitely clog servers with the right kind of a query. That is why this PR was created as this does recognize deeper levels and includes the necessary defined optimizations.

**Pull request merged**